## 들어가며

소프트 웨어 개발은 루프와 함수 말고도 알아야 할 내용이 상당히 많은데 초급 과정을 마치거나 프로그래밍 입문서를 읽은 뒤 더 많은 지침을 찾기 위해 또 다른 헬로 월드 튜토리얼로 이어지는 시기를 `절망의 사막`이라고 부른다. 이런 사막을 헤메는 사람들은 `가면 증후군`을 겪을 가능성이 많은데 자기가 `진짜` 프로그래머가 아니거나 `진짜` 프로그래머가 코딩 하는 방식을 알지 못한다고 느낀다.

## 추천 도서

- 나의 첫 파이썬 : 에릭 마테스
- 실전 파이썬 핸즈온 프로젝트 : 리 본(Lee Vaughan)
- 진지한 파이썬 : 쥘래앵 당주(Julien Danjou)
- 파이썬 코딩의 기술 : 브렛 슬라킨(Brett Slatkin)
- 파이썬 쿡북 : 다비드 비즐리(David Beazley)와 브라이언 코츠(Brian K. Jones)
- Fluent Python : 루시아누 램랄로(Luciano Ramalho)

# 1부 준비

## 1장 에러 메시지 파악과 고수의 도움을 받기 위한 에티켓

컴퓨터가 에러 메시지를 띄우는 것은 여러분이 컴퓨터를 불쾌하게 만들었기 때문이 아니다. 사람처럼 취급하지 말고 한낱 도구로 다뤄야 한다.

프로그래밍 학습은 대체로 독학 해야 하는데 파이썬을 공부하고도 하루에 몇 번씩 인터넷 검색을 하다보면 자신을 자책하게 된다. 하지만 전문 개발자들도 궁금한 점이 생기면 검색하거나 관련 문서를 참조하며 스스로 답을 찾는 습관은 알고리즘이나 데이터 구조 지식을 쌓는 일보다 훨씬 중요하다.

### 파이썬 에러 메시지를 읽는 법

대부분의 프로그래머는 전문 용어로 가득한 에러 메시지를 무시하고 싶은 충동이 든다. 하지만 에러 메시지에는 무엇이 잘못됐는지를 이해 할 수 있는 해답이 담겨있다.

답을 찾는 과정은 2단계로 먼저 traceback를 조사하고 에러 메시지를 인터넷에서 검색하는 것이다.

#### 추적정보 조사

파이썬에서는 except 문으로 처리할 수 없는 예외가 발생하면 crash를 낸다. 이 경우 에러 메시지와 추적 정보를 보여주는데 stack trace라고 불리는 이 정보는 예외가 발생한 위치와 그 예외로부터 이어진 함수 호출의 흔적을 보여준다.

```python
def a():
    print('Start of a()')
    b()

def b():
    print('Start of b()')
    c()

def c():
    print('Start of c()')
    5/0

a()
```

위의 코드를 실행하면 다음과 같은 결과를 얻는다.

```python
Start of a()
Start of b()
Start of c()
Traceback (most recent call last):
  File "test.py", line 13, in <module>
    a()
  File "test.py", line 3, in a
    b()
  File "test.py", line 7, in b
    c()
  File "test.py", line 11, in c
    5/0
ZeroDivisionError: division by zero
```

```python
Traceback (most recent call last)
```

다음 내용부터 추적 정보가 이어지며 첫째 함수 호출에서 시작해 가장 최근에 호출된 함수로 끝나는 것을 알려준다.

```python
File "test.py", line 13, in <module>
  a()
```

프로엠 요약으로 프레임 객체 내부의 정보를 보여준다. 함수가 호출되면 함수 호출이 끝난 후 반환될 코드 위치와 함께 지역 변수 데이터가 프레임 객체에 저장되고 함수가 호출될 때 생성되고 반환될 때 소멸된다.

추적정보에는 충돌로 이어진 각 프레임에 대한 프레임 요약이 표시되는데 13행에서 이 함수가 호출되었고 `<module>`이라는 문구로 이 행이 전역 범위에 있다는 사실을 알 수 있다.

```python
File "test.py", line 3, in a
  b()
File "test.py", line 7, in b
  c()
```

이어지는 행은 다음 두 프레임을 요약한다. print() 호출은 함수 호출 전에 실행됐지만 예외로 이어지는 함수 호출이 포함된 행만 표시된다.

마지막 프레임 요약에는 예외를 발생시킨 행이 표시되고 예외의 이름과 메시지가 표시된다.

```python
  File "test.py", line 11, in c
    5/0
ZeroDivisionError: division by zero
```

특정한 변수 값에서 division by zero 에러가 발생되는 경우 마지막 에러 메시지를 읽어도 문제를 확연히 알 수 없는 경우도 있고 파이썬 인터프리터가 다음 행을 읽기 전까지 구문 에러를 인지하지 못한 경우 다른 행을 가리킬 수 있다.

버그 파악을 위한 충분한 정보가 없거나 진짜 원인이 보이지 않는 이전 행에 있는 경우 디버거를 이용해 단계별로 따라가거나 로그를 확인해 원인을 찾아야 한다.

#### 에러 메시지 검색

에러 메시지는 온전한 설명을 하기 보다 무언가를 상기시키기 위한 목적이 크다. 해당 에러 메시지를 검색하면 의미와 가능성 높은 원인에 대해 쉽게 찾아 볼 수 있다.

개발 언어에 대해 발생한 모든 에러 메시지를 기억하는 건 불가능하다.

에러 메시지에 작성한 코드의 일부분이 포함된 경우 일부분을 삭제하고 검색하면 코드를 노출하지 않을 수 있고 검색에서 의미있는 결과가 없으면 전체 에러 메시지를 넣어 볼 수 있다.
